libs/datamodel/connectors/sql-datamodel-connector/src/mysql_datamodel_connector.rs

```diff
@@ -34,11 +115,145 @@ impl Connector for MySqlDatamodelConnector {
+        let native_type = match name {
+            INT_TYPE_NAME => MySqlType::Int,
+            SMALL_INT_TYPE_NAME => MySqlType::SmallInt,
+            TINY_INT_TYPE_NAME => MySqlType::TinyInt,
+            MEDIUM_INT_TYPE_NAME => MySqlType::MediumInt,
+            BIG_INT_TYPE_NAME => MySqlType::BigInt,
+            FLOAT_TYPE_NAME => MySqlType::Float,
+            DOUBLE_TYPE_NAME => MySqlType::Double,
+            CHAR_TYPE_NAME => {
+                if let Some(arg) = args.first() {
+                    MySqlType::Char(*arg)
+                } else {
+                    return Err(ConnectorError::new_argument_count_mismatch_error(CHAR_TYPE_NAME, 1, 0));
+                }
+            }
+            VAR_CHAR_TYPE_NAME => {
+                if let Some(arg) = args.first() {
+                    MySqlType::VarChar(*arg)
+                } else {
+                    return Err(ConnectorError::new_argument_count_mismatch_error(
+                        VAR_CHAR_TYPE_NAME,
+                        1,
+                        0,
+                    ));
+                }
+            }
+            TINY_TEXT_TYPE_NAME => MySqlType::TinyText,
+            TEXT_TYPE_NAME => MySqlType::Text,
+            MEDIUM_TEXT_TYPE_NAME => MySqlType::MediumText,
+            LONG_TEXT_TYPE_NAME => MySqlType::LongText,
+            DATE_TYPE_NAME => MySqlType::Date,
+            TIME_TYPE_NAME => {
+                if let Some(arg) = args.first() {
+                    MySqlType::Time(Option::from(*arg))
+                } else {
+                    MySqlType::Time(None)
+                }
+            }
+            DATETIME_TYPE_NAME => {
+                if let Some(arg) = args.first() {
+                    MySqlType::DateTime(Option::from(*arg))
+                } else {
+                    MySqlType::DateTime(None)
+                }
+            }
+            YEAR_TYPE_NAME => MySqlType::Year,
+            JSON_TYPE_NAME => MySqlType::JSON,
+
+            _ => unreachable!("This code is unreachable as the core must guarantee to just call with known names."),
+        };
+
+        let native_type_constructor = self.constructors.iter().find(|c| c.name.as_str() == name).unwrap();
+
+        if native_type_constructor._number_of_args != length
+            && native_type_constructor._number_of_optional_args != length
+        {
+            return Err(ConnectorError::new_argument_count_mismatch_error(
+                name,
+                native_type_constructor._number_of_args,
+                length,
+            ));
+        }
+
+        // check for compatability with scalar type
+        let compatable_prisma_scalar_type = native_type_constructor.prisma_type;
+        if compatable_prisma_scalar_type != scalar_type {
+            return Err(ConnectorError::new_incompatible_native_type_error(
+                "MySql",
+                scalar_type,
+                compatable_prisma_scalar_type,
+            ));
+        }
```

Comment:
I think you can do those checks in the `core` layer as well. Then a connector can assume that those invariants hold already. This would enable you to eliminate some duplication.